crates/containerd-shim-wasmtime/src/instance.rs

```diff
@@ -81,200 +54,208 @@ fn load_spec(bundle: String) -> Result<oci::Spec, Error> {
-            Ok(spec) => spec,
+        log::info!("deleting instance: {}", self.id);
+        match instance_exists(&self.rootdir, self.id.as_str()) {
+            Ok(exists) => {
+                if !exists {
+                    return Ok(());
+                }
+            }
             Err(err) => {
-                error!("Could not load spec, skipping cgroup cleanup: {}", err);
+                error!("could not find the container, skipping cleanup: {}", err);
                 return Ok(());
             }
-        };
-        let cg = oci::get_cgroup(&spec)?;
-        cg.delete()?;
+        }
+        let container_root = get_instance_root(&self.rootdir, self.id.as_str())?;
+        let container = Container::load(container_root).with_context(|| {
+            format!(
+                "could not load state for container {id}",
+                id = self.id.as_str()
+            )
+        });
+        match container {
+            Ok(mut container) => container.delete(true).map_err(|err| {
+                Error::Any(anyhow::anyhow!(
+                    "failed to delete container {}: {}",
+                    self.id,
+                    err
+                ))
+            })?,
+            Err(err) => {
+                error!("could not find the container, skipping cleanup: {}", err);
+                return Ok(());
+            }
+        }
+
         Ok(())
     }
 
     fn wait(&self, waiter: &Wait) -> Result<(), Error> {
+        log::info!("waiting for instance: {}", self.id);
         let code = self.exit_code.clone();
         waiter.set_up_exit_code_wait(code)
     }
 }
 
+impl Wasi {
+    fn build_container(
+        &self,
+        stdin: &str,
+        stdout: &str,
+        stderr: &str,
+        engine: Engine,
+    ) -> anyhow::Result<Container> {
+        let syscall = create_syscall();
+        let stdin = maybe_open_stdio(stdin).context("could not open stdin")?;
+        let stdout = maybe_open_stdio(stdout).context("could not open stdout")?;
+        let stderr = maybe_open_stdio(stderr).context("could not open stderr")?;
+
+        let wasmtime_executor = Box::new(WasmtimeExecutor { engine });
+
+        let default_executor = Box::<DefaultExecutor>::default();
+
+        if let Some(stdin) = stdin {
+            unsafe {
+                STDIN_FD = Some(dup(STDIN_FILENO));
+                dup2(stdin, STDIN_FILENO);
+            }
+        }
+        if let Some(stdout) = stdout {
+            unsafe {
+                STDOUT_FD = Some(dup(STDOUT_FILENO));
+                dup2(stdout, STDOUT_FILENO);
+            }
+        }
+        if let Some(stderr) = stderr {
+            unsafe {
+                STDERR_FD = Some(dup(STDERR_FILENO));
+                dup2(stderr, STDERR_FILENO);
+            }
+        }
+
+        let container = ContainerBuilder::new(self.id.clone(), syscall.as_ref())
+            .with_executor(vec![wasmtime_executor, default_executor])?
```

Comment:
could this have performance impact on linux containers?

Reply:
Which part do you care about? I think it's probably no problem.

Reply:
This comment was made after a conversation with my collegue, Simon, at Microsoft. He suggested that in a pod spec where it contains multiple containers, and one wasm container, the executor selection logic, which checks if a container is a wasm container or not, will run for every single containers. This might have a performance impact on the startup time.

Reply:
@Mossaka That is correct. Each check process should be short. However, in general, that process should be small enough to be acceptable for users. This is because most of the time during container startup from docker/kubelet/podman... is spent on the container image.

Reply:
What I'm thinking here is we could have 2 executors with the following order:

1. Read platform specific magic headers (Linux ELF / Windows PE)
2. Wasmtime exectuor

And in our wasmtime executor, assuming order depenendce in youki, we can assume anything that's not platform specific is wasm. Alternatively we can specifically inspect the wasm files (like we would need to do for ELF and PE) to check if they are wasm.

Reply:
What benefits do we get by putting linux / windows executor prior to wasmtime executor?

Reply:
Mostly so we don't need to check if something is wasm.

Reply:
Excuse my ignorance, but would it be possible to check the image platform early on?
If it's `.*/wasm(32)?` assume it's a wasm container, otherwise a linux container.

Reply:
That could work assuming we have the data... like in an annotation or something?

Reply:
Yeah, I think we could have a list of heuristics, going from explicit to implicit and quick to slow:
  1. Check the annotation -- if the user says this is a Wasm binary, believe them.
  2. Check the file extension -- if it's `.wasm`, consider it a Wasm binary.
  3. Check the ELF headers
  4. Maybe check Wasm binary format?

Note that possibly that in the (far) future we need to have information about which WASI interfaces are used by the module, so that we can use the right Wasm runtime with the right settings. It could be that we need to standardize some annotation for that purpose and we can use that for the binary check too. But that's not in the scope for this PR. :-)

Reply:
I am happy to keep this as is. We can evolve it in later PRs.
I also imagine this will change a fair bit when we start using OCI artifacts, and we can revisit this then.

Reply:
Yeah I prefer to keep it as is and I'd like to raise an issue for further discussion on how we can improve this. I'd like to get @jsturtevant thoughts on this as well.

Reply:
> That could work assuming we have the data... like in an annotation or something?

this could be done via an additional annotation short term in https://github.com/containerd/containerd/pull/8699. Longer term I could see a section in the [runtimeconfig](https://github.com/opencontainers/runtime-spec/blob/main/spec.md#platforms) for WASI (like it is for Windows/linux)

> I'd like to raise an issue for further discussion on how we can improve this

This sounds reasonable.  This is a good first step in that direction to get folks to try it out.

Reply:
So this PR made a decision to 
1. check Linux binary
2. check wasm binary

In checking Linux binary, it performs
1. arguments should not be empty
2. find binary from PATH or the path of the binary contains "/" (this follows youki's `verify_binary` impl)
3. check execute permission and existance of the file
4. check the shebang and ELF magic number

In checking wasm binary, it performs
1. check entrypoint path and truncate "#"
2. check extension is "wasm" or "wat"